This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [1]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = '##################'

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [2]:
# First, import the relevant modules
import requests

In [3]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
r = requests.get('https://www.quandl.com/api/v3/datasets/FSE/AFX_X/data.json?start_date=2016-12-31&end_date=2018-01-01&api_key=' + API_KEY)

In [4]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
data_json = r.json()

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [7]:
stock_dict=dict(data_json)
data_stock=stock_dict['dataset_data']['data']
columns = stock_dict['dataset_data']['column_names']
print(data_stock[-0])
print(columns)

['2017-12-29', 51.76, 51.94, 51.45, 51.76, None, 34640.0, 1792304.0, None, None, None]
['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover']


In [21]:
highest_open=0.0
lowest_open=99999999999.0
largest_daily_change=0.0
largest_change=0.0
close = None
trading_volume=0.0
counter=0.0
for i in data_stock:
    if i[2] is not None and i[3] is not None:
        daily_change = i[2]-i[3]
        if largest_daily_change < daily_change:
            largest_daily_change = daily_change
            date_daily_change = i[0]
    if i[1] is not None:
        if highest_open < i[1]:
            highest_open = i[1]
            date_highest = i[0]
        if lowest_open > i[1]:
            lowest_open = i[1]
            date_lowest = i[0]
    if close is not None:
        change = i[4] - close
        if largest_change < abs(change):
            largest_change = change
            date_change = i[0]
    close = i[4]
    trading_volume += i[6]
    counter += 1
ave_trading_volume = trading_volume/counter
    

In [23]:
print(largest_daily_change, date_daily_change)
print(highest_open, date_highest)
print(lowest_open, date_lowest)
print(largest_change, date_change)
print(ave_trading_volume)

2.8100000000000023 2017-05-11
53.11 2017-12-14
34.0 2017-01-24
2.559999999999995 2017-08-08
89124.33725490196


The highest the stock price was on December 14, 2017 at a price of 53.11.  The lowest the stock price was on January 24, 2017 at a price of 34.00.

The largest change on any one day was on May 11, 2017 at a price difference of approximately 2.81.

The largest change between two days happened on August 8, 2017 with a closing price difference of approximately 2.56.

The average trading volume was approximately 89124.